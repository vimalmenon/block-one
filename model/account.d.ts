export interface AccountParams {
  customerId: string
  accountNo: string
}

export interface AccountModel {
  customerId: number
  accountNumber: string
  accountType: string
  balance: number
}
