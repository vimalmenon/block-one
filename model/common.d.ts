export type GenericMethodType<K> = (value: K) => void
export type GenericReturnMethodType<K, V> = (value: K) => V

export interface TableColumn {
  header: string
  accessor: string
  template?: any
}
