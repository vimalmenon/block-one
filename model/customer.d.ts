export interface CustomerParams {
  customerId: string
}

export interface CustomerModel {
  prefix: string
  name: string
  customerId: number
  address: string
  emailAddress: string
  contactNo: string
}
