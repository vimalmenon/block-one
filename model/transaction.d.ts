export interface TransactionModel {
  accountNumber: string
  timestamp: string
  action: string
  description: string
  amount: number
  currency: 'HKD' | 'USD'
}
