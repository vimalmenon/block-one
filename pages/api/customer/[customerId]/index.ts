import type { NextApiRequest, NextApiResponse } from 'next'

import { openDb } from '../../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<any>) => {
  if (req.method === 'GET') {
    const { customerId } = req.query
    const result = await (
      await openDb()
    ).all(
      'SELECT prefix, name, customer_id as customerId, address as address, email_address as emailAddress,  contact_no as contactNo FROM customers WHERE customer_id = ? LIMIT 1',
      customerId
    )
    res.status(200).json(result[0])
  }
}
