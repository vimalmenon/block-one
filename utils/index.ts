export * from "./navigate";
export * from "./convertCurrency";