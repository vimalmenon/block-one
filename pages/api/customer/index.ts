import type { NextApiRequest, NextApiResponse } from 'next'
import { CustomerModel } from '../../../model/customer'

import { openDb } from '../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<CustomerModel[]>) => {
  if (req.method === 'GET') {
    const results = await (
      await openDb()
    ).all(
      'select prefix, name, customer_id as customerId, address, email_address as emailAddress, contact_no as contactNo from customers'
    )
    res.status(200).json(results)
  }
}
