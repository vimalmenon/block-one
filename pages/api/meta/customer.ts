import type { NextApiRequest, NextApiResponse } from 'next'
import { CustomerParams } from '../../../model/customer'

import { openDb } from '../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<CustomerParams[]>) => {
  if (req.method === 'GET') {
    const results = await (await openDb()).all('select customer_id as customerId FROM accounts')
    const params: CustomerParams[] = []
    results.forEach((result) => {
      params.push({
        customerId: result.customerId.toString(),
      })
    })
    res.status(200).json(params)
  }
}
