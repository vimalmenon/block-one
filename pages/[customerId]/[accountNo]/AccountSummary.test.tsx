import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { CustomerModel } from '../../../model/customer'
import { TransactionModel } from '../../../model/transaction'
import { onLinkNavigation } from '../../../utils'

import AccountSummary from './'

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: {
      customerId: '12345',
    },
  }),
}))

jest.mock('../../../utils', () => ({
  onLinkNavigation: jest.fn(),
}))

describe('AccountSummary Page', () => {
  test('AccountSummary: Snapshot testing', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const transactions: TransactionModel[] = [
      {
        accountNumber: '122-333-445',
        timestamp: '12 Dec 2020',
        action: 'Debit',
        description: 'Test data 1',
        amount: 20000,
        currency: 'HKD',
      },
      {
        accountNumber: '122-333-445',
        timestamp: '12 Dec 2020',
        action: 'Debit',
        description: 'Test data 2',
        amount: 25000,
        currency: 'HKD',
      },
    ]
    const wrapper = shallow(<AccountSummary customer={customer} transactions={transactions} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  test('AccountSummary: Back button testing', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const transactions: TransactionModel[] = [
      {
        accountNumber: '122-333-445',
        timestamp: '12 Dec 2020',
        action: 'Debit',
        description: 'Test data 1',
        amount: 20000,
        currency: 'HKD',
      },
      {
        accountNumber: '122-333-445',
        timestamp: '12 Dec 2020',
        action: 'Debit',
        description: 'Test data 2',
        amount: 25000,
        currency: 'HKD',
      },
    ]
    const wrapper = shallow(<AccountSummary customer={customer} transactions={transactions} />)
    wrapper.find('Button').simulate('click')
    expect(onLinkNavigation).toHaveBeenCalled()
    expect(onLinkNavigation).toHaveBeenCalledWith(`/12345`)
  })
})
