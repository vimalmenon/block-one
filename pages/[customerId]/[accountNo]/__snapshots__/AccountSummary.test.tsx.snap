// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSummary Page AccountSummary: Snapshot testing 1`] = `
<Container
  customer={
    Object {
      "address": "test Address",
      "contactNo": "123-456-789",
      "customerId": 12345,
      "emailAddress": "test.value@test.com",
      "name": "test value",
      "prefix": "Mr",
    }
  }
>
  <MetaData
    title="Transaction details for test value"
  />
  <div
    className="m-4 flex w-full flex-col py-10"
  >
    <Table
      columns={
        Array [
          Object {
            "accessor": "no",
            "header": "No",
            "template": [Function],
          },
          Object {
            "accessor": "timestamp",
            "header": "Timestamp",
          },
          Object {
            "accessor": "action",
            "header": "Action",
          },
          Object {
            "accessor": "description",
            "header": "Description",
          },
          Object {
            "accessor": "amount",
            "header": "Amount",
            "template": [Function],
          },
          Object {
            "accessor": "currency",
            "header": "Currency",
          },
        ]
      }
      items={
        Array [
          Object {
            "accountNumber": "122-333-445",
            "action": "Debit",
            "amount": 20000,
            "currency": "HKD",
            "description": "Test data 1",
            "timestamp": "12 Dec 2020",
          },
          Object {
            "accountNumber": "122-333-445",
            "action": "Debit",
            "amount": 25000,
            "currency": "HKD",
            "description": "Test data 2",
            "timestamp": "12 Dec 2020",
          },
        ]
      }
    />
    <div
      className="flex my-4 justify-between"
    >
      <div />
      <Button
        onClick={[Function]}
        type="secondary"
      >
        Back
      </Button>
    </div>
  </div>
</Container>
`;
